### DISCUSSION TITLE
proposal: constraints: new package to define standard type parameter constraints (discussion)

### DISCUSSION THREAD

**Discussion by rsc** [golang/go#47319] [Discussions]
We are ready to finish the discussion on #45458, because type sets have been accepted.
Here is the current API:

The intent is that if we add any new predeclared types, such as [`int128` and `uint128`](https://golang.org/issue/9455), those types will be supported by these constraints as appropriate based on the constraint descriptions.

```Go
// Package constraints defines a set of useful constraints to be used with type parameters.
package constraints

// Signed is a constraint that permits any signed integer type.
type Signed interface { ... }

// Unsigned is a constraint that permits any unsigned integer type.
type Unsigned interface { ... }

// Integer is a constraint that permits any integer type.
type Integer interface { ... }

// Float is a constraint that permits any floating-point type.
type Float interface { ... }

// Complex is a constraint that permits any complex numeric type.
type Complex interface { ... }

// Ordered is a constraint that permits any ordered type: any type that supports the operators < <= >= >.
type Ordered interface { ... }
```

**Comment by deanveloper** [comment]
[**Resolved**: Decided not to do this, at least not yet. - rsc, July 28 2021]
- - -

On the issue, I had mentioned that another useful type constraint would be "Arithmetic", which contained any type with `+, -, *, /` operators defined on them.  This would mean `Integer | Float | Complex`. An alternative name could be "Number", however that may be confusing because one may assume that "Number" is ordered when it is not.

**Reply by rsc** [reply]
This is easy to add in a future release, but we haven't seen a need for it yet (except in tests).

**Reply by rsc** [reply]
The general consensus is that Number is better than Arithmetic if we do this (but not yet if at all).

**Reply by rsc** [reply]
Resolved: decided not to do this for now.

**Comment by fzipp** [comment]
[**Resolved**: Decided to keep "constraints". - rsc, July 28 2021]
- - -

I'd prefer the singular form `constraint` as the package name, because in type parameter lists we refer to one constraint at a time: `constraint.Ordered`, `constraint.Integer`. It simply reads better in my opinion.

**Reply by ianlancetaylor** [reply]
On the other hand we already have packages named "bytes" and "strings", and there is a proposal for a package named "slices".

**Reply by fzipp** [reply]
I assumed "bytes", "strings" and "errors" were pluralized out of necessity to avoid collision with the respective builtin types, because all other package names in the standard library have names in singular form.

I don't mind "slices" as much as "constraints", because the proposed "slices" package has some contentual parallels to the "bytes" and "strings" packages. There the continued inconsistency may be justified.

On another note, "constraints" will be the second longest package name in the standard library after "quotedprintable". It will almost exclusively be used in type parameter lists, which are at the primary position of function signatures. This package name will visually dominate a function signature it is in, and move everything else a fair amount to the right side of the line, even though it conveys little information: we already know something must be a constraint if it is used in constraint position. So a name for this package should be chosen with care, and optics should be taken into consideration.

**Reply by earthboundkid** [reply]
What about “isa”? `isa.Integer`, etc.

**Reply by earthboundkid** [reply]
One more (because “is a unsigned” sounds weird): types. Types fits the existing Go pattern of name plus S equals package for dealing with name.

**Reply by jimmyfrasche** [reply]
If we want to go super short I'd think `is` would suffice and probably be fine since this only gets used in declarations

**Reply by earthboundkid** [reply]
I thought about `is`, but that’s already the name of a testing library (not that the conflict matters per se). I really like `types` though. It feels very “Go” to me.

**Reply by deanveloper** [reply]
I like “types” a lot more. Also - It’s not that I don’t like “constraints” for its length (even though I dont _particularly_ like how long it is). However, I’d like to erase the difference between a “constraint” and an ordinary interface.

**Reply by jimmyfrasche** [reply]
There's already go/types

**Reply by rsc** [reply]
Technically this _could_ be "constraint"
We've got bytes, strings, errors, and so on because those cannot use the singular.
maps will not be able to use the singular either.
slices could, but having to remember "maps.Equal" vs "slice.Equal" is a bit weird.
It seems more consistent in this case to just use plural for all of these helping-builtins packages, even when they don't strictly need to be (slices, constraints).

**Reply by rsc** [reply]
Resolved: "constraints" is the color of this bike shed.

**Comment by benjaminjkraft** [comment]
[**Resolved**: Will make docs clear and add a vet check. - rsc, July 28 2021]
- - -

Is adding types in the future going to cause compatibility problems?  (This is reposting and merging two of my comments on the issue.)

A simple but perhaps too contrived example:
```go
type MySigned interface { type int8, int16, int32, int64 }
func MyAbs[T MySigned](v T) T { ... }
func Abs[T constraints.Signed](v T) T { return MyAbs(v) }
```

(Of course presumably `Abs` is in some other package, and perhaps the type lists involved are more complex.) That compiles today, but not in a future Go version where `constraints.Signed` contains `int128`.

A perhaps more likely example (assuming #45346 is accepted):

```go
func Abs[T constraints.Float](v T) v {
  switch T {
  case ~float32:
    return math.Float32FromBits(math.Float32Bits(v) &^ (1 << 31))
  case ~float64:
    return math.Float64FromBits(math.Float64Bits(v) &^ (1 << 63)) // math.Abs
  default:
    panic("unknown type")
  }
}
```

(I'm using syntax proposed in #45380 here for brevity, but since `T` cannot be of interface type I believe this code can be expressed equally well without it, as `switch any(v).(type)` + `case interface{ ~float32 }`.) Right now this can never panic; if `float128` is added it can, although at least it still compiles.

It doesn't seem like this should block the addition of the `constraints` package, but I don't see an obvious way to work around it.  Part of the problem here is that experienced Go programmers know what it takes to make existing Go types backwards-compatible: for structs it's fine to add methods and fields (assuming no unkeyed literals); for interfaces may add methods only if they have an unexported method; for functions you just can't change them. But I, at least, don't know those rules for type lists yet -- these examples suggest that, more like interfaces, they can't be changed without some (in this case unknown) setup.

**Reply by deanveloper** [reply]
It won't cause any existing code to break, however it will will not work with any _new_ usages of `Abs`. I don't think this is within the scope of the compatibility promise.

**Reply by benjaminjkraft** [reply]
To be clear, in the first example `Abs` would fail to compile.  In the second, yes, it would only fail if actually passed an `int128`.

**Reply by deanveloper** [reply]
Ahh I see the issue in the first example now. That's a good point. I think something like that may be a good candidate for something like Go module versioning.

**Reply by Merovius** [reply]
> Is adding types in the future going to cause compatibility problems? (This is reposting and merging two of my comments on the issue.)

I think so. There is a simple example:

```go
import "constraints"

type MySigned interface {
    int8 | int16 | int32 | int64
}

func F[A MySigned]() {
    G[A]()
}

func G[B constraint.Signed]() {
    F[B]()
}
```

Of course this example isn't doing anything useful, but it illustrates the problem. If you add `int128` to `constraint.Signed`, the instantiation of `F` in `G` will fail - `B` might be `int128`, which is not in `MySigned`. Likewise, if you'd remove a type (say `int64`) from `constraint.Signed`, the instantiation of `G` in `F` will fail, as `A` might be `uint64`.

As long as you don't mix `constraint` constraints with your own, this won't be a problem. I'm not sure that will be an easy restriction to follow or communicate.

[edit] ah, I see that's your first example m( I only looked at the second.

**Reply by bcmills** [reply]
At the very least, it seems like we should add a point or two about type constraints to https://golang.org/doc/go1compat#expectations.

(The situation seems loosely analogous to the addition of fields to struct literals, or the addition of methods to types.)

**Reply by Merovius** [reply]
@bcmills I disagree about this being analog to fields to struct literals or methods to types.

Adding fields to struct literals is an easy fix without downsides (except a bit more verbosity), while it's not necessarily easy to avoid mixing `constraints`-based constraints and custom ones - in any case, it might be a change in semantics, not just syntax.

Adding methods on types is just generally not a problem. It might sometimes change semantics, when combined with embedding, but generally, it's really safe.

Removing/Adding types to a type-set is analogous to adding/removing methods to an interface, respectively. It has exactly the same failure modes. Except we don't even have the workaround-mechanism of "optional interfaces" (AFAICT).

We might have to just bite that bullet. But it's not nice. In particular, the problematic case is if generic code that uses a custom type-set is instantiating generic code that uses an stdlib constraint (reasoning that the custom constraint covers the relevant type-set). But *most* generic code (including the ones in the stdlib) will likely use stdlib constraints. So this feels akin to telling people to effectively avoid using generic library code from generic code - if they want to use custom constraints. That feels unsavory to me.

**Reply by bcmills** [reply]
> Adding methods on types is just generally not a problem. It might sometimes change semantics, when combined with embedding, but generally, it's really safe.

When combined with embedding,  adding methods causes _severe_ problems. (Consider what happens if you add a `String` or `MarshalJSON` method to a field embedded within a struct that does not itself define those methods.) So, the compatibility rule for struct types is: if you want to wrap a type whose method set may be expanded, _never_ embed an instance of that type in your own struct.

> Removing/Adding types to a type-set is analogous to adding/removing methods to an interface, respectively. It has exactly the same failure modes.

Correct. And it is dual to adding methods to a concrete type, and has a dual compatibility rule: if you want to implement an interface whose method set may be expanded, _always_ embed an instance of that interface in your own struct.

Union elements in interfaces are dual to interface embedding, so the same rule applies here with the directions swapped. If you want to ensure that `constraints.Signed` always implements `MySigned`, then `constraints.Signed` must be included in each union element of `MySigned`.

This rule is mechanical and fairly straightfoward to enforce in code review. It's not obvious to me whether it is simple enough for a `vet` check, though.

**Reply by bcmills** [reply]
> In particular, the problematic case is if generic code that uses a custom type-set is instantiating generic code that uses an stdlib constraint (reasoning that the custom constraint covers the relevant type-set). But most generic code (including the ones in the stdlib) will likely use stdlib constraints.

Most generic functions in the standard library will be _invoked by_ user code, but will not _invoke_ user code itself. So that constraint will go in the easier direction: any type that implements `constraints.Signed` today will continue to implement it tomorrow without any explicit intervention, and so any user code that calls a standard-library function will continue to be able to call it going forward.

**Reply by bcmills** [reply]
In other words, in your example it is realistic for `F` to call `G`, but it would be unusual for `G` to call `F`.

**Reply by Merovius** [reply]
I think you are right, I got turned around with which case is the harder one.

**Reply by ianlancetaylor** [reply]
My sense is that this is a real issue, but it is one that must be addressed by documentation and style guidelines rather than by changing the proposed package.

Does anybody disagree with that?

**Reply by Merovius** [reply]
@ianlancetaylor I think leaving it to documentation and style guidelines assumes that there *is* a good recommendation we can make. So, yes, changing this package won't be the solution either way. But the question is still relevant as it may point to a larger problem with having this package and/or our approach to generics.

That being said, I'm inclined to agree that it *can* be solved in documentation, given the discussion above.

**Reply by ianlancetaylor** [reply]
I think the style guideline is, roughly: 1) given a type parameter whose constraint is taken from the constraints package, don't use that type parameter to instantiate a type/function whose constraint is not `any` and is not taken from the constraints package; 2) don't write a type switch to handle all types when using a constraint taken from the constraints package.

I think this is the kind of rough spot that is hard to avoid when using type sets (or type lists) to determine operation sets.  It's not theoretically clean, but, since new types that support operations are added very rarely, it's manageable in practice.

**Reply by benjaminjkraft** [reply]
Yeah.  I think condition (1) can get less obvious to a human if there are a few more layers involved: instead of `func Abs[T constraints.Signed](v T) T { return MyAbs(v) }` say you have

```go
type SignedAndIsValid interface {
  constraints.Signed
  IsValid() bool
}
func MyValidAbs[T SignedAndIsValid](v T) T {
  if v.IsValid() { return MyAbs(v) }
  return v
}
```

and now it's using a `SignedAndIsValid` as `MySigned` that you don't want to do, even though you defined both.  But it does still seem vet-checkable.

(Also, I think we may want to suggest that for any third-party package, not just `constraints`?  Since they might embed `constraints` in which case the same issue applies, or just want to similarly add a new type.  The rule might get a little more complex since if the third-party constraint embeds `constraint.Signed`, probably it's safe for you to use that as a `constraint.Signed`, just not as a `MySigned`.  Still vet-checkable, I would think.)

**Reply by Merovius** [reply]
@benjaminjkraft I think for the purpose of the recommendation, only the type-set part of the interfaces is relevant - i.e. your `SignedAndIsValid`, for the purpose of this rule, is a `constraint`-package based constraint.

**Reply by benjaminjkraft** [reply]
Yeah, of course this example is no more "interesting".  The point is it's less obvious to a human: you aren't using the `constraints` package directly at all when you define `MyValidAbs`.  You have to know that you are using it indirectly.

**Reply by rsc** [reply]
It sounds like we know what the rules need to be, and that we can write a vet check for them. That should be enough to keep people from accidentally writing code that would not be "int128-safe".

**Reply by rsc** [reply]
Resolved: will make sure we have a vet check for the expectations around these integer constraints.

**Reply by balasanjay** [reply]
Out of curiosity, could the constraints package declare a local unexported type (maybe called `futureInts` or similar), and include that in the constraint?

That would ensure that attempting to assign `constraints.Signed` to any externally declared constraint would presumably not work.

(This is similar to the idea of writing interfaces with an unexported method)

**Reply by ianlancetaylor** [reply]
My thinking is that a local unexported type would work, but I worry that it would be a cryptic wart for most future readers, to solve a problem that I think is quite unlikely and that we can in any case check for in other ways.

**Comment by changkun** [comment]
I have a question about `Ordered`. If there is a self-defined composite literal that does not support operators `<` `<=` `>=` `>` but methods like `Less` `LessEqual`, then how to let the type constrained in `Ordered` so that we can utilize any standard package which parameters are constrained using `constraints.Ordered`?

**Reply by Merovius** [reply]
It is not possible. A generic function can *either* be written use parameters *or* methods, but never both. If you want to support both predeclared types and composite types, you need to write two functions.

**Comment by rsc** [comment]
**Update, July 28 2021**

Based on the discussion, we will make the following changes to the proposed API:

- Added constraints.Slice to support more readable slice type parameters, based on discussion in #47203.
- Will make forward-compatibility requirements for Integer etc clear in docs and add vet check.
- Decided to keep "constraints" as the name of the package.
- Decided against Number (not enough evidence of need).

I have also added [**Resolved**] notes at the top of comments that we consider resolved, in addition to replying with text like "Resolved...". 

@ianlancetaylor is going to update the proposal text.

**Comment by Merovius** [comment]
Given that we are adding `constraints.Slice[Elem]`, would it also make sense to add `constraints.Map[Key, Value]` for #47330?

**Reply by earthboundkid** [reply]
I was in the middle of typing this:

> Following on from constraints.Slice, what about constraints.Map? It would be useful e.g. for dealing with http.Header and url.Values using the maps package.

**Reply by rsc** [reply]
Talked to @ianlancetaylor, and it seems like we should add Slice, Map, and Chan.

**Reply by ianlancetaylor** [reply]
Now done, over a #45458.

**Comment by rsc** [comment]
**Update, August 4 2021**

Based on the discussion, we will make the following changes to the proposed API:

- Add constraints.Map and constraints.Chan to match constraints.Slice.

@ianlancetaylor is going to update the proposal text.
It seems like we are getting close to likely accept. Perhaps next week.

I am also starting to wonder how long it will take me to learn to type the word constraints.

**Reply by urandom** [reply]
Shouldn't the Key type parameter for Map be ordered?

**Reply by Merovius** [reply]
@urandom Why would it? You can use `type T struct{ x int }` as a map-key, even though it is not ordered.

**Reply by Merovius** [reply]
Oh, perhaps you mean `comparable`. It probably should. Though I wouldn't expect it to make much of a difference in practice, as you just can't have a type which satisfies that constraint, unless `K` is also `comparable` (i.e. the `comparable` is kind of implied). But it would likely be clearer if it was explicit.

**Reply by deanveloper** [reply]
Under the idea the proposal, wouldn't `type Map[Key, Val any] interface { map[Key]Val }` fail to compile?

**Reply by Merovius** [reply]
I'm not certain. The go2go tools fails to compile it, but I would argue there is really no reason. The resulting type set would just be empty, unless `K` is comparable. But generally, we allow to write down constraints, even if they result in empty type sets.

**Reply by ianlancetaylor** [reply]
The `Key` type for `Map` should be `comparable`.  Fixed.

**Reply by ianlancetaylor** [reply]
@deanveloper Sorry, not quite seeing what you are getting at.  Did you mean to add a "constraints." somewhere in there?

**Reply by ianlancetaylor** [reply]
@deanveloper Oh, I see.  Yes, I think that code like

    type M[K any] map[K]int

should fail to compile.  (It would work if it were `K comparable`.)  Based on that, probably your example should also fail to compile.

**Reply by deanveloper** [reply]
@ianlancetaylor I was mentioning the previous state of `constraints.Map`, wondering if using `any` for the map key would fail to compile.

@Merovius If that were the case, would I be able to write a constraint like `type Foo[T any] interface { ~StringsOnly[T] }` (where `StringsOnly` is something like `type StringsOnly[T interface{ ~string }] interface { Bar(T) T }`)? That doesn't seem like a good idea.

**Reply by Merovius** [reply]
@deanveloper I don't understand your example. I do think its type-set would be empty (`StringsOnly[T]` is a defined type and there are no types whose underlying type is a defined type - except the predeclared types), but I don't understand what it would be intended to do.

Either way, the problem is that a general proof of a type-set being empty is difficult. [I wrote down a proof of that here](https://github.com/golang/go/issues/45346#issuecomment-813639125). It might be possible [under the latest restrictions](https://github.com/golang/go/issues/45346#issuecomment-862505803). I haven't checked, but the restrictions where put in place to solve very similar problems, so it's not unlikely they would make it possible to determine if a type set is empty as well. Someone would have to sit down and write down an algorithm to do so (or a proof that we still can't) and we would then have to decide if that algorithm is sufficiently simple to end up in the spec - or sufficiently self-evident that we don't have to.

Meanwhile, there are no real downsides to allowing constraints even if their type set is empty. Even the most casual of tests (i.e. even just trying to call a function using that constraint) will immediately surface the problem, as that constraint can't be instantiated, so the function can't be called.

Lastly: This is probably not the place to have this discussion. Note that the immediate concern - making the type-parameter in `Map` comparable - is already addressed.

**Reply by urandom** [reply]
@Merovius right, I meant comparable. Late night comments arent well testable

**Comment by Merovius** [comment]
Presumably, the `Slice`, `Map` and `Chan` constraints should use approximation-elements? Otherwise we wouldn't much need them :)

**Reply by ianlancetaylor** [reply]
Yes, done.  Thanks.

**Comment by Merovius** [comment]
I believe most functions accepting channels should restrict the direction of communication. The `Chan` constraint as is, though, doesn't allow that - you have to choose between writing `func F[E any](ch <-chan E)`, which excludes defined types, or write `func F[C Chan[E], E any](ch C)`, which doesn't verify directionality.

We could add `ReadOnlyChan` and `WriteOnlyChan`. But even then, could `func F[C ReadOnlyChan[E], E any](ch C)` be instantiated with a plain `chan T`, or would it have to be explicitly converted to `<-chan T`? I'm genuinely asking, FWIW, I can't really think of the exact rules and how they would apply to this.
